File: ios/lib/storage.ts
import AsyncStorage from '@react-native-async-storage/async-storage';
import { MONITORED_APPS } from './constants';

// --- Types ---

export interface IntentionEntry {
  app: string;
  intention: string;
  timestamp: number;
  proceeded: boolean; // true = continued to app, false = went back
}

export interface SessionEntry {
  app: string;
  timestamp: number;
  duration: number; // breathing duration in seconds
  proceeded: boolean;
}

export interface Settings {
  monitoredApps: string[];         // app IDs from MONITORED_APPS
  breathingDuration: number;       // seconds (6, 15, or 30)
  hapticsEnabled: boolean;
  onboardingComplete: boolean;
}

export interface Stats {
  interventionCount: number;
  turnaroundCount: number;         // times user chose "go back"
  totalBreathingSeconds: number;
  currentStreak: number;           // consecutive days
  lastSessionDate: string | null;  // ISO date string (YYYY-MM-DD)
  sessions: SessionEntry[];
  intentionLog: IntentionEntry[];
}

// --- Keys ---

const KEYS = {
  settings: 'dsf_settings',
  stats: 'dsf_stats',
};

// --- Defaults ---

const DEFAULT_SETTINGS: Settings = {
  monitoredApps: MONITORED_APPS.map(a => a.id),
  breathingDuration: 6,
  hapticsEnabled: true,
  onboardingComplete: false,
};

const DEFAULT_STATS: Stats = {
  interventionCount: 0,
  turnaroundCount: 0,
  totalBreathingSeconds: 0,
  currentStreak: 0,
  lastSessionDate: null,
  sessions: [],
  intentionLog: [],
};

// --- Settings ---

export async function getSettings(): Promise<Settings> {
  try {
    const raw = await AsyncStorage.getItem(KEYS.settings);
    if (!raw) return { ...DEFAULT_SETTINGS };
    return { ...DEFAULT_SETTINGS, ...JSON.parse(raw) };
  } catch {
    return { ...DEFAULT_SETTINGS };
  }
}

export async function updateSettings(partial: Partial<Settings>): Promise<Settings> {
  const current = await getSettings();
  const updated = { ...current, ...partial };
  await AsyncStorage.setItem(KEYS.settings, JSON.stringify(updated));
  return updated;
}

// --- Stats ---

export async function getStats(): Promise<Stats> {
  try {
    const raw = await AsyncStorage.getItem(KEYS.stats);
    if (!raw) return { ...DEFAULT_STATS };
    return { ...DEFAULT_STATS, ...JSON.parse(raw) };
  } catch {
    return { ...DEFAULT_STATS };
  }
}

export async function recordSession(
  app: string,
  duration: number,
  proceeded: boolean,
  intention: string
): Promise<void> {
  const stats = await getStats();
  const now = Date.now();
  const today = new Date().toISOString().split('T')[0];

  // Update counts
  stats.interventionCount += 1;
  if (!proceeded) stats.turnaroundCount += 1;
  stats.totalBreathingSeconds += duration;

  // Update streak
  if (stats.lastSessionDate === today) {
    // Already logged today, streak unchanged
  } else if (stats.lastSessionDate === getYesterday()) {
    stats.currentStreak += 1;
  } else {
    stats.currentStreak = 1;
  }
  stats.lastSessionDate = today;

  // Add session
  stats.sessions.push({ app, timestamp: now, duration, proceeded });
  // Keep last 200 sessions
  if (stats.sessions.length > 200) {
    stats.sessions = stats.sessions.slice(-200);
  }

  // Add intention
  if (intention.trim()) {
    stats.intentionLog.push({ app, intention, timestamp: now, proceeded });
    // Keep last 50 intentions
    if (stats.intentionLog.length > 50) {
      stats.intentionLog = stats.intentionLog.slice(-50);
    }
  }

  await AsyncStorage.setItem(KEYS.stats, JSON.stringify(stats));
}

export async function resetStats(): Promise<void> {
  await AsyncStorage.setItem(KEYS.stats, JSON.stringify({ ...DEFAULT_STATS }));
}

export async function resetAll(): Promise<void> {
  await AsyncStorage.multiRemove([KEYS.settings, KEYS.stats]);
}

// --- Helpers ---

function getYesterday(): string {
  const d = new Date();
  d.setDate(d.getDate() - 1);
  return d.toISOString().split('T')[0];
}

export function getTodaySessions(sessions: SessionEntry[]): SessionEntry[] {
  const today = new Date().toISOString().split('T')[0];
  return sessions.filter(s => {
    const d = new Date(s.timestamp).toISOString().split('T')[0];
    return d === today;
  });
}

export function getWeekSessions(sessions: SessionEntry[]): SessionEntry[] {
  const weekAgo = Date.now() - 7 * 24 * 60 * 60 * 1000;
  return sessions.filter(s => s.timestamp >= weekAgo);
}

File: xcode/Doomscrollingfix/Doomscrollingfix/Storage.swift
import Foundation

struct SessionEntry: Codable, Identifiable {
    var id: Double { timestamp }
    let app: String
    let timestamp: Double
    let duration: Double
    let proceeded: Bool
}

struct IntentionEntry: Codable, Identifiable {
    var id: Double { timestamp }
    let app: String
    let intention: String
    let timestamp: Double
    let proceeded: Bool
}

struct AppSettings: Codable {
    var monitoredApps: [String]
    var breathingDuration: Int
    var hapticsEnabled: Bool
    var onboardingComplete: Bool

    static var `default`: AppSettings {
        AppSettings(
            monitoredApps: allMonitoredAppIds,
            breathingDuration: 6,
            hapticsEnabled: true,
            onboardingComplete: false
        )
    }
}

struct AppStats: Codable {
    var interventionCount: Int
    var turnaroundCount: Int
    var totalBreathingSeconds: Double
    var currentStreak: Int
    var lastSessionDate: String?
    var sessions: [SessionEntry]
    var intentionLog: [IntentionEntry]

    static let empty = AppStats(
        interventionCount: 0,
        turnaroundCount: 0,
        totalBreathingSeconds: 0,
        currentStreak: 0,
        lastSessionDate: nil,
        sessions: [],
        intentionLog: []
    )
}

@Observable
final class AppStore {
    private let settingsKey = "dsf_settings"
    private let statsKey = "dsf_stats"
    private let decoder = JSONDecoder()
    private let encoder = JSONEncoder()

    var settings: AppSettings
    var stats: AppStats

    init() {
        if let data = UserDefaults.standard.data(forKey: "dsf_settings"),
           let s = try? JSONDecoder().decode(AppSettings.self, from: data) {
            self.settings = s
        } else {
            self.settings = .default
        }
        if let data = UserDefaults.standard.data(forKey: "dsf_stats"),
           let s = try? JSONDecoder().decode(AppStats.self, from: data) {
            self.stats = s
        } else {
            self.stats = .empty
        }
    }

    func saveSettings() {
        if let data = try? encoder.encode(settings) {
            UserDefaults.standard.set(data, forKey: settingsKey)
        }
    }

    func saveStats() {
        if let data = try? encoder.encode(stats) {
            UserDefaults.standard.set(data, forKey: statsKey)
        }
    }

    func recordSession(app: String, duration: Double, proceeded: Bool, intention: String) {
        let now = Date().timeIntervalSince1970 * 1000
        let today = Self.todayString()

        stats.interventionCount += 1
        if !proceeded { stats.turnaroundCount += 1 }
        stats.totalBreathingSeconds += duration

        if stats.lastSessionDate == today {
            // already logged today
        } else if stats.lastSessionDate == Self.yesterdayString() {
            stats.currentStreak += 1
        } else {
            stats.currentStreak = 1
        }
        stats.lastSessionDate = today

        stats.sessions.append(SessionEntry(
            app: app, timestamp: now, duration: duration, proceeded: proceeded
        ))
        if stats.sessions.count > 200 {
            stats.sessions = Array(stats.sessions.suffix(200))
        }

        if !intention.trimmingCharacters(in: .whitespaces).isEmpty {
            stats.intentionLog.append(IntentionEntry(
                app: app, intention: intention, timestamp: now, proceeded: proceeded
            ))
            if stats.intentionLog.count > 50 {
                stats.intentionLog = Array(stats.intentionLog.suffix(50))
            }
        }

        saveStats()
    }

    func resetStats() {
        stats = .empty
        saveStats()
    }

    var todaySessions: [SessionEntry] {
        let today = Self.todayString()
        return stats.sessions.filter { entry in
            let d = Date(timeIntervalSince1970: entry.timestamp / 1000)
            return Self.dateString(d) == today
        }
    }

    var weekSessions: [SessionEntry] {
        let weekAgo = Date().timeIntervalSince1970 * 1000 - 7 * 24 * 60 * 60 * 1000
        return stats.sessions.filter { $0.timestamp >= weekAgo }
    }

    var turnaroundRate: Int {
        guard stats.interventionCount > 0 else { return 0 }
        return Int(round(Double(stats.turnaroundCount) / Double(stats.interventionCount) * 100))
    }

    var estimatedTimeSaved: Int {
        stats.turnaroundCount * 5
    }

    private static func todayString() -> String {
        dateString(Date())
    }

    private static func yesterdayString() -> String {
        dateString(Calendar.current.date(byAdding: .day, value: -1, to: Date())!)
    }

    private static func dateString(_ date: Date) -> String {
        let f = DateFormatter()
        f.dateFormat = "yyyy-MM-dd"
        f.timeZone = .current
        return f.string(from: date)
    }

    static func formatDuration(_ seconds: Double) -> String {
        let s = Int(seconds)
        if s < 60 { return "\(s)s" }
        let m = s / 60
        let rem = s % 60
        return rem > 0 ? "\(m)m \(rem)s" : "\(m)m"
    }
}
